src/zh/cartoon18/src/eu/kanade/tachiyomi/extension/zh/cartoon18/Cartoon18.kt

```diff
@@ -0,0 +1,233 @@
+        launchIO { fetchKeywords() }
+        return FilterList(
+            SortFilter(),
+            if (keywordsList.isEmpty()) {
+                Filter.Header("Tap 'Reset' to load keywords")
+            } else {
+                KeywordFilter(keywordsList)
+            },
+        )
+    }
+
+    private open class QueryFilter(
+        name: String,
+        values: Array<String>,
+        private val queryName: String,
+        private val queryValues: Array<String>,
+        state: Int = 0,
+    ) : Filter.Select<String>(name, values, state) {
+        fun addQueryTo(builder: HttpUrl.Builder) =
+            builder.addQueryParameter(queryName, queryValues[state])
+    }
+
+    private class SortFilter : QueryFilter(
+        "Sort by",
+        arrayOf("Latest", "Popular", "Recommended", "Best"),
+        "sort",
+        arrayOf("created", "hits", "score", "likes"),
+        state = 2,
+    )
+
+    class Keyword(val name: String, val value: String)
+
+    private var keywordsList: List<Keyword> = emptyList()
+
+    private class KeywordFilter(keywords: List<Keyword>) : QueryFilter(
+        "Keyword",
+        keywords.map { it.name }.toTypedArray(),
+        "q",
+        keywords.map { it.value }.toTypedArray(),
+    )
+
+    /**
+     * Inner variable to control how much tries the keywords request was called.
+     */
+    private var fetchKeywordsAttempts: Int = 0
+
+    /**
+     * Fetch the keywords from the source to be used in the filters.
+     */
+    private fun fetchKeywords() {
+        if (fetchKeywordsAttempts < 3 && keywordsList.isEmpty()) {
+            try {
+                keywordsList = client.newCall(keywordsRequest()).execute()
+                    .use { parseKeywords(it.asJsoup()) }
+            } catch (_: Exception) {
+            } finally {
+                fetchKeywordsAttempts++
+            }
+        }
+    }
+
+    /**
+     * The request to the search page (or another one) that have the keywords list.
+     */
+    private fun keywordsRequest(): Request {
+        return GET("$baseUrlWithLang/category".toHttpUrl(), headers)
+    }
+
+    /**
+     * Get the keywords from the search page document.
+     *
+     * @param document The search page document
+     */
+    private fun parseKeywords(document: Document): List<Keyword> {
+        val items = document.select("div.content a.btn")
+        return buildList(items.size + 1) {
+            add(Keyword("None", ""))
+            items.mapTo(this) {
+                val value = it.text()
+                val queryValue = URLDecoder.decode(it.attr("href").substringAfterLast('/'), "UTF-8")
+                Keyword(value, queryValue)
+            }
+        }
+    }
+
+    private val scope = CoroutineScope(Dispatchers.IO)
+
+    private fun launchIO(block: () -> Unit) = scope.launch { block() }
+
+    private val preferences by lazy {
+        Injekt.get<Application>().getSharedPreferences("source_$id", 0x0000)!!
+    }
```

Comment:
since you are using the preference immediately, remove the lazy
```suggestion
    private val preferences =
        Injekt.get<Application>().getSharedPreferences("source_$id", 0x0000)
```